File: src/cpp/src/pipeline/resonance.cpp
#include "molgr/pipeline/resonance.h"

#include <deque>
#include <map>
#include <memory>
#include <queue>
#include <set>
#include <tuple>
#include <utility>

namespace
{
    struct QueueEntry
    {
        int discrepancy = 0;
        int depth = 0;
        int order = 0;
        std::shared_ptr<OpenBabel::OBMol> omol;
        molgr::reconstruct::ResonanceStateKey state_key;
    };

    struct QueueEntryGreater
    {
        bool operator()(const QueueEntry &lhs, const QueueEntry &rhs) const
        {
            return std::tie(lhs.discrepancy, lhs.depth, lhs.order) >
                   std::tie(rhs.discrepancy, rhs.depth, rhs.order);
        }
    };
}

namespace molgr
{
    namespace reconstruct
    {
        void WalkRadicalResonances(
            const OpenBabel::OBMol &mol,
            int max_depth,
            ResonanceVisitCallback visit)
        {
            if (!visit)
            {
                visit = [](const ResonanceSearchNode &) { return true; };
            }

            const auto [root_key, bond_index_map] = BuildResonanceSearchContext(mol);
            std::set<ResonanceStateKey> seen{root_key};
            std::deque<std::tuple<std::shared_ptr<OpenBabel::OBMol>, ResonanceStateKey, int>> frontier{
                {std::make_shared<OpenBabel::OBMol>(mol), root_key, 0}};

            while (!frontier.empty())
            {
                auto [current, current_key, depth] = std::move(frontier.front());
                frontier.pop_front();

                const bool should_expand = visit(ResonanceSearchNode{
                    *current,
                    current_key,
                    depth,
                });
                if (depth >= max_depth || !should_expand)
                {
                    continue;
                }

                const auto moves = EnumerateOneStepResonanceMoves(*current, current_key, bond_index_map);
                for (const auto &move : moves)
                {
                    if (seen.find(move.next_state_key) != seen.end())
                    {
                        continue;
                    }
                    seen.insert(move.next_state_key);
                    frontier.emplace_back(
                        MaterializeOneStepResonancePtr(*current, move.idxs),
                        move.next_state_key,
                        depth + 1);
                }
            }
        }

        void WalkRadicalResonancesLimitedDiscrepancy(
            const OpenBabel::OBMol &mol,
            int max_depth,
            ResonanceVisitCallback visit,
            const LimitedDiscrepancyTraversalConfig &traversal_config,
            const molgr::config::MolGRConfig &config)
        {
            if (!visit)
            {
                visit = [](const ResonanceSearchNode &) { return true; };
            }

            const auto [root_key, bond_index_map] = BuildResonanceSearchContext(mol);
            std::map<ResonanceStateKey, int> best_discrepancy_by_state{{root_key, 0}};
            std::set<ResonanceStateKey> emitted_states;
            std::priority_queue<QueueEntry, std::vector<QueueEntry>, QueueEntryGreater> frontier;
            frontier.push(QueueEntry{0, 0, 0, std::make_shared<OpenBabel::OBMol>(mol), root_key});

            int push_order = 0;
            while (!frontier.empty())
            {
                QueueEntry current_entry = std::move(frontier.top());
                frontier.pop();

                const auto best_it = best_discrepancy_by_state.find(current_entry.state_key);
                if (best_it == best_discrepancy_by_state.end() ||
                    current_entry.discrepancy != best_it->second)
                {
                    continue;
                }

                bool should_expand = true;
                if (emitted_states.find(current_entry.state_key) == emitted_states.end())
                {
                    emitted_states.insert(current_entry.state_key);
                    should_expand = visit(ResonanceSearchNode{
                        *current_entry.omol,
                        current_entry.state_key,
                        current_entry.depth,
                    });
                }

                if (current_entry.depth >= max_depth || !should_expand)
                {
                    continue;
                }

                const auto moves =
                    EnumerateOneStepResonanceMoves(*current_entry.omol, current_entry.state_key, bond_index_map);
                const auto selected_moves =
                    SelectLimitedDiscrepancyMoves(*current_entry.omol, moves, traversal_config, config);
                for (std::size_t move_rank = 0; move_rank < selected_moves.size(); ++move_rank)
                {
                    const int next_discrepancy =
                        current_entry.discrepancy + static_cast<int>(move_rank);
                    if (next_discrepancy > traversal_config.max_discrepancy)
                    {
                        break;
                    }

                    const auto &move = selected_moves[move_rank];
                    const auto best_known_it = best_discrepancy_by_state.find(move.next_state_key);
                    if (best_known_it != best_discrepancy_by_state.end() &&
                        best_known_it->second <= next_discrepancy)
                    {
                        continue;
                    }

                    best_discrepancy_by_state[move.next_state_key] = next_discrepancy;
                    frontier.push(QueueEntry{
                        next_discrepancy,
                        current_entry.depth + 1,
                        ++push_order,
                        MaterializeOneStepResonancePtr(*current_entry.omol, move.idxs),
                        move.next_state_key,
                    });
                }
            }
        }

        std::vector<OpenBabel::OBMol> GetRadicalResonances(const OpenBabel::OBMol &mol, int max_depth)
        {
            std::vector<OpenBabel::OBMol> resonances;
            WalkRadicalResonances(
                mol,
                max_depth,
                [&resonances](const ResonanceSearchNode &node)
                {
                    resonances.push_back(node.omol);
                    return true;
                });
            return resonances;
        }

        std::vector<OpenBabel::OBMol> GetRadicalResonancesLimitedDiscrepancy(
            const OpenBabel::OBMol &mol,
            int max_depth,
            const LimitedDiscrepancyTraversalConfig &traversal_config,
            const molgr::config::MolGRConfig &config)
        {
            std::vector<OpenBabel::OBMol> resonances;
            WalkRadicalResonancesLimitedDiscrepancy(
                mol,
                max_depth,
                [&resonances](const ResonanceSearchNode &node)
                {
                    resonances.push_back(node.omol);
                    return true;
                },
                traversal_config,
                config);
            return resonances;
        }
    }
}

File: src/molgr/fallback/pipeline/resonance.py
"""Radical resonance traversal flow for the fallback pipeline."""

from __future__ import annotations

from collections import deque
from heapq import heappop, heappush
from typing import Callable, Deque, List, Optional, Tuple

from openbabel import pybel

from molgr.fallback.utils import resonance as resonance_utils
from molgr.fallback.utils.resonance import (
    ProcessedResonanceKey,
    ResonanceBondIndexMap,
    ResonanceSearchNode,
    ResonanceStateKey,
    ResonanceTraversalContext,
    ResonanceTraversalMove,
    ResonanceTraversalPolicy,
    build_processed_resonance_key,
    build_resonance_state_key,
    make_limited_discrepancy_direct_gain_traversal_policy,
    make_limited_discrepancy_force_field_traversal_policy,
    make_limited_discrepancy_input_order_traversal_policy,
    process_resonance,
    resonance_move_score_cache_clear,
    resonance_move_score_cache_info,
)


def get_radical_resonances(
    omol: pybel.Molecule,
    max_depth: int = 2,
    *,
    traversal_policy: Optional[ResonanceTraversalPolicy] = None,
) -> List[pybel.Molecule]:
    """Collect raw resonance states reachable within the configured search depth."""

    resonances: List[pybel.Molecule] = []

    def _collect(node: ResonanceSearchNode) -> bool:
        resonances.append(node.omol)
        return True

    walk_radical_resonances(
        omol,
        max_depth=max_depth,
        traversal_policy=traversal_policy,
        visit=_collect,
    )
    return resonances


def walk_radical_resonances(
    omol: pybel.Molecule,
    max_depth: int = 2,
    *,
    traversal_policy: Optional[ResonanceTraversalPolicy] = None,
    visit: Optional[Callable[[ResonanceSearchNode], bool]] = None,
) -> None:
    """Traverse resonance states and let the caller decide whether each node should expand."""

    if isinstance(traversal_policy, resonance_utils._LIMITED_DISCREPANCY_POLICY_TYPES):
        _walk_radical_resonances_limited_discrepancy(
            omol,
            max_depth=max_depth,
            traversal_policy=traversal_policy,
            visit=visit,
        )
        return

    _walk_radical_resonances_bfs(
        omol,
        max_depth=max_depth,
        traversal_policy=traversal_policy,
        visit=visit,
    )


def _walk_radical_resonances_bfs(
    omol: pybel.Molecule,
    *,
    max_depth: int,
    traversal_policy: Optional[ResonanceTraversalPolicy],
    visit: Optional[Callable[[ResonanceSearchNode], bool]],
) -> None:
    visitor = visit if visit is not None else (lambda _node: True)
    root_key, bond_index_map = resonance_utils._build_resonance_search_context(omol)
    seen = {root_key}
    frontier: Deque[Tuple[pybel.Molecule, ResonanceStateKey, int]] = deque([(omol, root_key, 0)])

    while frontier:
        current, current_key, depth = frontier.popleft()
        should_expand = visitor(
            ResonanceSearchNode(
                omol=current,
                state_key=current_key,
                depth=depth,
            )
        )
        if depth >= max_depth or not should_expand:
            continue

        indexed_moves = resonance_utils._enumerate_one_step_resonance_moves(
            current,
            current_key,
            bond_index_map,
        )
        selected_moves = resonance_utils._apply_resonance_traversal_policy(
            ResonanceTraversalContext(
                root_omol=omol,
                current_omol=current,
                current_state_key=current_key,
                depth=depth,
                max_depth=max_depth,
            ),
            indexed_moves,
            traversal_policy,
        )
        for move in selected_moves:
            if move.next_state_key in seen:
                continue
            seen.add(move.next_state_key)
            frontier.append(
                (
                    resonance_utils._materialize_one_step_resonance(current, move.idxs),
                    move.next_state_key,
                    depth + 1,
                )
            )


def _walk_radical_resonances_limited_discrepancy(
    omol: pybel.Molecule,
    *,
    max_depth: int,
    traversal_policy: ResonanceTraversalPolicy,
    visit: Optional[Callable[[ResonanceSearchNode], bool]],
) -> None:
    visitor = visit if visit is not None else (lambda _node: True)
    root_key, bond_index_map = resonance_utils._build_resonance_search_context(omol)
    best_discrepancy_by_state = {root_key: 0}
    emitted_states = set()
    frontier: List[Tuple[int, int, int, pybel.Molecule, ResonanceStateKey]] = [
        (0, 0, 0, omol, root_key)
    ]
    push_order = 0

    while frontier:
        discrepancy, depth, _order, current, current_key = heappop(frontier)
        if discrepancy != best_discrepancy_by_state.get(current_key):
            continue
        if current_key in emitted_states:
            should_expand = True
        else:
            emitted_states.add(current_key)
            should_expand = visitor(
                ResonanceSearchNode(
                    omol=current,
                    state_key=current_key,
                    depth=depth,
                )
            )
        if depth >= max_depth or not should_expand:
            continue

        indexed_moves = resonance_utils._enumerate_one_step_resonance_moves(
            current,
            current_key,
            bond_index_map,
        )
        selected_moves = resonance_utils._apply_resonance_traversal_policy(
            ResonanceTraversalContext(
                root_omol=omol,
                current_omol=current,
                current_state_key=current_key,
                depth=depth,
                max_depth=max_depth,
            ),
            indexed_moves,
            traversal_policy,
        )
        for move_rank, move in enumerate(selected_moves):
            next_discrepancy = discrepancy + move_rank
            if next_discrepancy > traversal_policy.max_discrepancy:
                break
            best_known_discrepancy = best_discrepancy_by_state.get(move.next_state_key)
            if best_known_discrepancy is not None and best_known_discrepancy <= next_discrepancy:
                continue
            best_discrepancy_by_state[move.next_state_key] = next_discrepancy
            push_order += 1
            heappush(
                frontier,
                (
                    next_discrepancy,
                    depth + 1,
                    push_order,
                    resonance_utils._materialize_one_step_resonance(current, move.idxs),
                    move.next_state_key,
                ),
            )


__all__ = [
    "ProcessedResonanceKey",
    "ResonanceBondIndexMap",
    "ResonanceSearchNode",
    "ResonanceStateKey",
    "ResonanceTraversalContext",
    "ResonanceTraversalMove",
    "ResonanceTraversalPolicy",
    "build_processed_resonance_key",
    "build_resonance_state_key",
    "get_radical_resonances",
    "make_limited_discrepancy_direct_gain_traversal_policy",
    "make_limited_discrepancy_force_field_traversal_policy",
    "make_limited_discrepancy_input_order_traversal_policy",
    "process_resonance",
    "resonance_move_score_cache_clear",
    "resonance_move_score_cache_info",
    "walk_radical_resonances",
]
